File: typescript/src/core/WebhookRouter.ts
import { IncomingMessage, ServerResponse } from 'http';
import {
  WebhookEvent,
  WebhookProvider,
  WebhookRouterConfig,
  WebhookHandler,
  HandlerRegistry
} from '../types';
import {
  StripeProvider,
  GitHubProvider,
  SlackProvider,
  TwilioProvider,
  SquareProvider
} from '../providers';
import { Logger, getDefaultLogger } from '../utils/logger';

export class WebhookRouter {
  private config: WebhookRouterConfig;
  private providers: Map<string, WebhookProvider> = new Map();
  private handlers: HandlerRegistry = {};
  private customProviders: Map<string, WebhookProvider> = new Map();
  private logger: Logger;
  
  constructor(config: WebhookRouterConfig, logger?: Logger) {
    this.config = config;
    this.logger = logger || getDefaultLogger();
    this.initializeProviders();
  }
  
  private initializeProviders(): void {
    // Register built-in providers
    if (this.config.stripe) {
      this.providers.set('stripe', StripeProvider);
    }
    if (this.config.github) {
      this.providers.set('github', GitHubProvider);
    }
    if (this.config.slack) {
      this.providers.set('slack', SlackProvider);
    }
    if (this.config.twilio) {
      this.providers.set('twilio', TwilioProvider);
    }
    if (this.config.square) {
      this.providers.set('square', SquareProvider);
    }
  }
  
  public on(provider: string, eventType: string, handler: WebhookHandler): void {
    if (!this.handlers[provider]) {
      this.handlers[provider] = {};
    }
    this.handlers[provider][eventType] = handler;
  }
  
  public registerProvider(provider: WebhookProvider): void {
    this.customProviders.set(provider.name, provider);
  }
  
  public async handle(req: IncomingMessage, res: ServerResponse): Promise<void> {
    try {
      // Extract headers and body
      const headers = this.normalizeHeaders(req.headers);
      const rawBody = await this.getRawBody(req);
      
      this.logger.debug('Processing webhook request', { 
        headers: Object.keys(headers).join(', '),
        bodySize: rawBody.length 
      });
      
      // Identify provider
      const provider = this.identifyProvider(headers, rawBody);
      if (!provider) {
        this.logger.warn('Unknown webhook source', { headers });
        res.statusCode = 400;
        res.end('Unknown webhook source');
        return;
      }
      
      this.logger.debug(`Identified provider: ${provider.name}`);
      
      // Get provider config
      const providerConfig = this.config[provider.name];
      if (!providerConfig) {
        res.statusCode = 400;
        res.end('Provider not configured');
        return;
      }
      
      // Verify signature
      const isValid = provider.verify(headers, rawBody, providerConfig);
      if (!isValid) {
        this.logger.warn(`Invalid signature for ${provider.name} webhook`);
        res.statusCode = 403;
        res.end('Invalid signature');
        return;
      }
      
      this.logger.debug(`Signature verified for ${provider.name}`);
      
      // Parse payload
      let payload: any;
      try {
        payload = provider.parsePayload(rawBody, headers);
      } catch (error) {
        res.statusCode = 400;
        res.end('Invalid payload');
        return;
      }
      
      // Handle special cases
      if (provider.name === 'slack' && payload.type === 'url_verification') {
        res.statusCode = 200;
        res.setHeader('Content-Type', 'text/plain');
        res.end(payload.challenge);
        return;
      }
      
      // Extract event type
      const eventType = provider.extractEventType(headers, payload);
      
      // Create event object
      const event: WebhookEvent = {
        provider: provider.name,
        type: eventType,
        id: this.extractEventId(provider.name, headers, payload),
        payload,
        rawHeaders: headers,
        rawBody: rawBody.toString('utf8'),
        receivedAt: new Date()
      };
      
      // Find and execute handler
      const handler = this.findHandler(provider.name, eventType);
      
      if (handler) {
        try {
          this.logger.info(`Executing handler for ${provider.name}/${eventType}`, {
            eventId: event.id
          });
          
          const result = await handler(event);
          
          // Handle response
          if (result !== undefined && result !== null) {
            res.statusCode = 200;
            if (typeof result === 'object') {
              res.setHeader('Content-Type', 'application/json');
              res.end(JSON.stringify(result));
            } else {
              res.end(String(result));
            }
          } else {
            res.statusCode = 200;
            res.end();
          }
          
          this.logger.info(`Successfully processed ${provider.name}/${eventType}`, {
            eventId: event.id
          });
        } catch (error) {
          this.logger.error(`Handler error for ${provider.name}/${eventType}`, error as Error, {
            provider: provider.name,
            eventType,
            eventId: event.id
          });
          res.statusCode = 500;
          res.end('Internal server error');
        }
      } else {
        // No handler found, but webhook is valid
        res.statusCode = 200;
        res.end();
      }
    } catch (error) {
      this.logger.error('Webhook processing error', error as Error);
      res.statusCode = 500;
      res.end('Internal server error');
    }
  }
  
  private normalizeHeaders(headers: IncomingMessage['headers']): Record<string, string> {
    const normalized: Record<string, string> = {};
    for (const [key, value] of Object.entries(headers)) {
      normalized[key.toLowerCase()] = Array.isArray(value) ? value[0] : value || '';
    }
    return normalized;
  }
  
  private async getRawBody(req: IncomingMessage): Promise<Buffer> {
    return new Promise((resolve, reject) => {
      const chunks: Buffer[] = [];
      req.on('data', (chunk) => chunks.push(chunk));
      req.on('end', () => resolve(Buffer.concat(chunks)));
      req.on('error', reject);
    });
  }
  
  private identifyProvider(headers: Record<string, string>, body: Buffer): WebhookProvider | null {
    // Check built-in providers first
    for (const provider of this.providers.values()) {
      if (provider.identify(headers, body)) {
        return provider;
      }
    }
    
    // Check custom providers
    for (const provider of this.customProviders.values()) {
      if (provider.identify(headers, body)) {
        return provider;
      }
    }
    
    return null;
  }
  
  private findHandler(provider: string, eventType: string): WebhookHandler | null {
    const providerHandlers = this.handlers[provider];
    if (!providerHandlers) {
      return null;
    }
    
    // Check for exact match
    if (providerHandlers[eventType]) {
      return providerHandlers[eventType];
    }
    
    // Check for wildcard
    if (providerHandlers['*']) {
      return providerHandlers['*'];
    }
    
    return null;
  }
  
  private extractEventId(provider: string, headers: Record<string, string>, payload: any): string | undefined {
    switch (provider) {
      case 'stripe':
        return payload.id;
      case 'github':
        return headers['x-github-delivery'];
      case 'slack':
        return payload.event_id || payload.event?.event_id;
      default:
        return undefined;
    }
  }
}

File: python/unified_webhook_router/core/webhook_router.py
import asyncio
import inspect
from datetime import datetime
from typing import Dict, Any, Optional, Callable, Union
from ..types import WebhookEvent, WebhookHandler, WebhookProvider, InvalidWebhookError
from ..providers import (
    StripeProvider,
    GitHubProvider,
    SlackProvider,
    TwilioProvider,
    SquareProvider
)
from ..utils.logger import WebhookLogger, NoOpLogger, get_default_logger


class WebhookRouter:
    def __init__(self, config: Dict[str, Any], logger: Optional[Union[WebhookLogger, NoOpLogger]] = None):
        self.config = config
        self.providers: Dict[str, WebhookProvider] = {}
        self.handlers: Dict[str, Dict[str, WebhookHandler]] = {}
        self.custom_providers: Dict[str, WebhookProvider] = {}
        self.logger = logger or get_default_logger()
        self._initialize_providers()
    
    def _initialize_providers(self):
        # Register built-in providers
        if 'stripe' in self.config:
            self.providers['stripe'] = StripeProvider()
        if 'github' in self.config:
            self.providers['github'] = GitHubProvider()
        if 'slack' in self.config:
            self.providers['slack'] = SlackProvider()
        if 'twilio' in self.config:
            self.providers['twilio'] = TwilioProvider()
        if 'square' in self.config:
            self.providers['square'] = SquareProvider()
    
    def on(self, provider: str, event_type: str, handler: Optional[WebhookHandler] = None):
        """Register a webhook handler. Can be used as a decorator or method."""
        def decorator(handler_func: WebhookHandler) -> WebhookHandler:
            if provider not in self.handlers:
                self.handlers[provider] = {}
            self.handlers[provider][event_type] = handler_func
            return handler_func
        
        if handler is None:
            # Used as decorator
            return decorator
        else:
            # Used as method
            decorator(handler)
    
    def register_provider(self, provider: WebhookProvider):
        """Register a custom webhook provider."""
        self.custom_providers[provider.name] = provider
    
    async def handle_request(self, request) -> Optional[Any]:
        """Handle a webhook request. Framework-agnostic method."""
        # Extract headers and body
        headers = self._normalize_headers(self._get_headers(request))
        raw_body = await self._get_raw_body(request)
        
        self.logger.debug('Processing webhook request', {
            'headers': ', '.join(headers.keys()),
            'body_size': len(raw_body)
        })
        
        # Identify provider
        provider = self._identify_provider(headers, raw_body)
        if not provider:
            self.logger.warn('Unknown webhook source', {'headers': headers})
            raise InvalidWebhookError('Unknown webhook source')
        
        self.logger.debug(f'Identified provider: {provider.name}')
        
        # Get provider config
        provider_config = self.config.get(provider.name)
        if not provider_config:
            raise InvalidWebhookError(f'Provider {provider.name} not configured')
        
        # Verify signature
        if not provider.verify(headers, raw_body, provider_config):
            self.logger.warn(f'Invalid signature for {provider.name} webhook')
            raise InvalidWebhookError('Invalid signature')
        
        self.logger.debug(f'Signature verified for {provider.name}')
        
        # Parse payload
        try:
            payload = provider.parse_payload(raw_body, headers)
        except Exception as e:
            raise InvalidWebhookError(f'Invalid payload: {str(e)}')
        
        # Handle special cases
        if provider.name == 'slack' and payload.get('type') == 'url_verification':
            return payload.get('challenge')
        
        # Extract event type
        event_type = provider.extract_event_type(headers, payload)
        
        # Create event object
        event = WebhookEvent(
            provider=provider.name,
            type=event_type,
            id=self._extract_event_id(provider.name, headers, payload),
            payload=payload,
            raw_headers=headers,
            raw_body=raw_body.decode('utf-8'),
            received_at=datetime.now()
        )
        
        # Find and execute handler
        handler = self._find_handler(provider.name, event_type)
        
        if handler:
            try:
                self.logger.info(f'Executing handler for {provider.name}/{event_type}', {
                    'event_id': event.id
                })
                
                # Execute handler
                if inspect.iscoroutinefunction(handler):
                    result = await handler(event)
                else:
                    result = handler(event)
                
                self.logger.info(f'Successfully processed {provider.name}/{event_type}', {
                    'event_id': event.id
                })
                
                return result
            except Exception as e:
                self.logger.error(f'Handler error for {provider.name}/{event_type}', e, {
                    'provider': provider.name,
                    'event_type': event_type,
                    'event_id': event.id
                })
                raise
        else:
            self.logger.debug(f'No handler registered for {provider.name}/{event_type}')
        
        # No handler found, but webhook is valid
        return None
    
    def _normalize_headers(self, headers: Dict[str, Any]) -> Dict[str, str]:
        """Normalize headers to lowercase string keys."""
        normalized = {}
        for key, value in headers.items():
            normalized[key.lower()] = str(value) if not isinstance(value, list) else str(value[0])
        return normalized
    
    def _get_headers(self, request) -> Dict[str, Any]:
        """Extract headers from various request objects."""
        # Flask/Werkzeug
        if hasattr(request, 'headers'):
            return dict(request.headers)
        # Django
        elif hasattr(request, 'META'):
            headers = {}
            for key, value in request.META.items():
                if key.startswith('HTTP_'):
                    header_name = key[5:].replace('_', '-').lower()
                    headers[header_name] = value
                elif key in ('CONTENT_TYPE', 'CONTENT_LENGTH'):
                    headers[key.replace('_', '-').lower()] = value
            return headers
        # FastAPI/Starlette
        elif hasattr(request, 'headers'):
            return dict(request.headers)
        else:
            raise ValueError('Unable to extract headers from request object')
    
    async def _get_raw_body(self, request) -> bytes:
        """Extract raw body from various request objects."""
        # Flask/Werkzeug
        if hasattr(request, 'get_data'):
            return request.get_data()
        # Django
        elif hasattr(request, 'body'):
            return request.body
        # FastAPI/Starlette (async)
        elif hasattr(request, 'body') and inspect.iscoroutinefunction(request.body):
            return await request.body()
        # AIOHTTP
        elif hasattr(request, 'read'):
            return await request.read()
        else:
            raise ValueError('Unable to extract body from request object')
    
    def _identify_provider(self, headers: Dict[str, str], body: bytes) -> Optional[WebhookProvider]:
        """Identify which provider sent the webhook."""
        # Check built-in providers first
        for provider in self.providers.values():
            if provider.identify(headers, body):
                return provider
        
        # Check custom providers
        for provider in self.custom_providers.values():
            if provider.identify(headers, body):
                return provider
        
        return None
    
    def _find_handler(self, provider: str, event_type: str) -> Optional[WebhookHandler]:
        """Find the appropriate handler for an event."""
        provider_handlers = self.handlers.get(provider)
        if not provider_handlers:
            return None
        
        # Check for exact match
        if event_type in provider_handlers:
            return provider_handlers[event_type]
        
        # Check for wildcard
        if '*' in provider_handlers:
            return provider_handlers['*']
        
        return None
    
    def _extract_event_id(self, provider: str, headers: Dict[str, str], payload: Any) -> Optional[str]:
        """Extract event ID if available."""
        if provider == 'stripe':
            return payload.get('id')
        elif provider == 'github':
            return headers.get('x-github-delivery')
        elif provider == 'slack':
            return payload.get('event_id') or (payload.get('event', {}).get('event_id'))
        return None